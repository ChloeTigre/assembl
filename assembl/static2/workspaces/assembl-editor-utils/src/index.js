// @flow
import { EditorState, Modifier, RichUtils } from 'draft-js';
import DraftJSPluginsUtils from 'draft-js-plugins-utils';

import { getSelectedBlockKey } from './blocks';
import * as constants from './constants';
import { getEntityRange } from './entities';
import { createSelectionState } from './selection';

type LinkData = {
  target: ?string,
  text: ?string,
  title: ?string,
  url: string
};

const { collapseToEnd, getCurrentEntity, getCurrentEntityKey, hasEntity, removeLinkAtSelection } = DraftJSPluginsUtils;

const { ENTITY_TYPES, ENTITY_MUTABILITY } = constants;

export default {
  collapseToEnd: collapseToEnd,

  createLinkAtSelection: (editorState: EditorState, data: LinkData): EditorState => {
    const { text } = data;
    const contentState = editorState.getCurrentContent().createEntity(ENTITY_TYPES.link, ENTITY_MUTABILITY.mutable, data);
    const entityKey = contentState.getLastCreatedEntityKey();
    let withLink;
    if (text) {
      const newContentState = Modifier.replaceText(contentState, editorState.getSelection(), text, undefined, entityKey);
      withLink = EditorState.createWithContent(newContentState);
    } else {
      withLink = RichUtils.toggleLink(editorState, editorState.getSelection(), entityKey);
    }
    return EditorState.forceSelection(withLink, editorState.getSelection());
  },

  replaceLinkAtCursor: (editorState: EditorState, data: LinkData): EditorState => {
    const { text } = data;
    const contentState = editorState.getCurrentContent();
    const blockKey = getSelectedBlockKey(editorState);
    const replacedEntityKey = getCurrentEntityKey(editorState);
    if (replacedEntityKey) {
      const entityRange = getEntityRange(editorState, replacedEntityKey);
      if (text && entityRange) {
        const { start, end } = entityRange;
        const replacedSelection = createSelectionState(blockKey, start, end);
        const newContentState = Modifier.replaceText(contentState, replacedSelection, text, null, replacedEntityKey);
        newContentState.mergeEntityData(replacedEntityKey, data);
        return EditorState.createWithContent(newContentState);
      }
    }

    return editorState;
  },

  getCurrentEntityKey: getCurrentEntityKey,

  getCurrentEntity: getCurrentEntity,

  hasEntity: hasEntity,

  removeLinkAtSelection: removeLinkAtSelection
};

export { constants, createSelectionState };