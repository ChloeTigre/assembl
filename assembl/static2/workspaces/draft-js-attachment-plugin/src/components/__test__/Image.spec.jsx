// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { AtomicBlockUtils, ContentState, EditorState } from 'draft-js';

import { constants } from 'assembl-editor-utils';

import Image from '../Image';

const { ENTITY_MUTABILITY, ENTITY_TYPES } = constants;

function mockFileReader(img) {
  const addEventListener = jest.fn((_, evtHandler) => {
    evtHandler();
  });
  const readAsDataURLSpy = jest.fn();
  const DummyFileReader = {
    addEventListener: addEventListener,
    readAsDataURL: readAsDataURLSpy,
    result: img
  };
  global.FileReader = jest.fn(() => DummyFileReader);
  return DummyFileReader;
}

describe('Image component', () => {
  it('should render an image (url)', () => {
    let contentState = ContentState.createFromText('');
    // $FlowFixMe DraftEntityType is too restrictive in DraftJS (see https://github.com/facebook/draft-js/issues/868 )
    contentState = contentState.createEntity(ENTITY_TYPES.image, ENTITY_MUTABILITY.immutable, {
      mimeType: 'image/png',
      src: 'my-img.png',
      title: 'My image'
    });
    const entityKey = contentState.getLastCreatedEntityKey();
    let editorState = EditorState.createWithContent(contentState);
    editorState = AtomicBlockUtils.insertAtomicBlock(editorState, entityKey, ' ');
    contentState = editorState.getCurrentContent();
    const blocks = contentState.getBlocksAsArray();
    const atomicBlock = blocks[1];

    const props = {
      block: atomicBlock,
      contentState: contentState
    };
    const component = renderer.create(<Image {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an image (File)', () => {
    let contentState = ContentState.createFromText('');
    const img = window.btoa('gimme some base64');
    // $FlowFixMe DraftEntityType is too restrictive in DraftJS (see https://github.com/facebook/draft-js/issues/868 )
    const src = new File([img], 'foobar.jpg');
    const DummyFileReader = mockFileReader(img);
    contentState = contentState.createEntity(ENTITY_TYPES.image, ENTITY_MUTABILITY.immutable, {
      mimeType: 'image/png',
      src: src,
      title: 'My image'
    });
    const entityKey = contentState.getLastCreatedEntityKey();
    let editorState = EditorState.createWithContent(contentState);
    editorState = AtomicBlockUtils.insertAtomicBlock(editorState, entityKey, ' ');
    contentState = editorState.getCurrentContent();
    const blocks = contentState.getBlocksAsArray();
    const atomicBlock = blocks[1];

    const props = {
      block: atomicBlock,
      contentState: contentState
    };
    const component = renderer.create(<Image {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(DummyFileReader.readAsDataURL).toHaveBeenCalledWith(src);
  });
});