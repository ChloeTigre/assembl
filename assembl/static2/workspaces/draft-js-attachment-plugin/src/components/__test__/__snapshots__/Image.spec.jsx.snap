// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component should render an image (File) 1`] = `
<div
  className="atomic-block"
  data-blocktype="atomic"
>
  <img
    alt="foobar.jpg"
    className="attachment-image"
    src="Z2ltbWUgc29tZSBiYXNlNjQ="
    title="foobar.jpg"
  />
</div>
`;

exports[`Image component should render an image (url) 1`] = `
<div
  className="atomic-block"
  data-blocktype="atomic"
>
  <img
    alt="My image"
    className="attachment-image"
    src="my-img.png"
    title="My image"
  />
</div>
`;
